core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs

```diff
@@ -1849,6 +1849,24 @@ writeNWith alloc n = FL.take n (writeNWithUnsafe alloc n)
 writeN :: forall m a. (MonadIO m, Storable a) => Int -> Fold m a (Array a)
 writeN = writeNWith newArray
 
+-- /Pre-release/
+{-# INLINE writeRevN #-}
+writeRevN :: forall m a. (MonadIO m, Storable a) => Int -> Fold m a (Array a)
+writeRevN n = Fold step initial (return . fst)
+    where
+
+    initial = do
+        arr <- newArray (max n 0)
+        return $ FL.Partial (arr, n - 1)
+
+    step (arr, i) x =
+        if i < 0
+        then
+            return $ FL.Done arr
+        else do
+            putIndex i x arr
+            return $ FL.Partial (arr, i - 1)
```

Comment:
Can you write this in the same way as `writeN` has been written? Using an array in the state would thread around an additional variable which might cause trouble with optimization. Also the tuple in the state is not strict.